ocaml/xapi/xapi_pif.ml

```diff
@@ -105,6 +105,39 @@ let read_bridges_from_inventory () =
   with _ ->
     []
 
+(* Ensure the PIF is not a bond slave. *)
+let assert_not_bond_slave_of ~__context ~self =
+  if Db.PIF.get_bond_slave_of ~__context ~self <> Ref.null then
+    raise (Api_errors.Server_error (Api_errors.cannot_plug_bond_slave, [Ref.string_of self]))
+
+(* Ensure the PIF has valid IPv4 configuration mode. *)
+let assert_valid_ip_configuration ~__context ~self =
+  if Db.PIF.get_ip_configuration_mode ~__context ~self = `None then
+    raise (Api_errors.Server_error(Api_errors.pif_has_no_network_configuration, [Ref.string_of self]))
+
+(* Ensure the PIF has valid IPv6 configuration mode. *)
+let assert_valid_ipv6_configuration ~__context ~self =
+  if Db.PIF.get_ipv6_configuration_mode ~__context ~self = `None then
+    raise (Api_errors.Server_error(Api_errors.pif_has_no_v6_network_configuration, [Ref.string_of self]))
+
+let assert_usable_for_management ~__context ~for_mgmt ~required_address_type ~self =
+  assert_not_bond_slave_of ~__context ~self;
+  let primary_address_type = Db.PIF.get_primary_address_type ~__context ~self in
+  if for_mgmt && required_address_type <> primary_address_type then
+    raise (Api_errors.Server_error(Api_errors.pif_incompatible_primary_address_type, [Ref.string_of self]));
+  if Db.PIF.get_managed ~__context ~self = true then begin
+    match primary_address_type with
+      | `IPv4 -> assert_valid_ip_configuration ~__context ~self 
+      | `IPv6 -> assert_valid_ipv6_configuration ~__context ~self 
+      | _ -> try
+        let mgmt_pif = Helpers.get_management_interface ~__context ~host:(Helpers.get_localhost ~__context) in
```

Comment:
Ah, that makes sense as behaviour. I don't think the original code did that though. A mistake? The original code is (paraphrasing slightly)
```
if pif is ipv4 and doesn't have a proper ipv4 configuration then
    raise an exception
else if pif is ipv6 and doesn't have a proper ipv6 configuration then
    raise an exception
else try
    ...
    if primary_address_type <> mgmt_address_type then
          raise (Api_errors.Server_error(Api_errors.pif_incompatible_primary_address_type, []));
    with _ ->
        () (* no current management interface *)
```

Note In particular the last `else` in the snippet above, which is in the line copied from https://github.com/xenserver/xen-api/pull/266/files/5e1756034dc254405d6bde6224972ec6f588a6d1#diff-126a0619d5403393ed49475b5dd96aa4L807
(Also it seems odd that we raise an error then immediately catch it.)